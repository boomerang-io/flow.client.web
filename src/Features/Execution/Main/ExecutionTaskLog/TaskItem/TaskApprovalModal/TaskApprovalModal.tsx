import React from "react";
import { queryCache, useMutation } from "react-query";
// import moment from "moment";
import { Formik } from "formik";
import * as Yup from "yup";
import {
  Button,
  DecisionButtons,
  InlineNotification,
  Loading,
  ModalBody,
  ModalForm,
  ModalFooter,
  notify,
  TextArea,
  ToastNotification,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { ThumbsUp16, ThumbsDown16 } from "@carbon/icons-react";
import styles from "./taskApprovalModal.module.scss";

const GateStatus = {
  Approved: "APPROVED",
  Rejected: "REJECTED",
};

type Props = {
  approvalId: string;
  executionId: string;
  closeModal: () => void;
};

function TaskApprovalModal({ approvalId, executionId, closeModal }: Props) {
  const cancelRequestRef = React.useRef<any>();

  const [approvalMutator, { isLoading: approvalsIsLoading, error: approvalsError }] = useMutation(
    (args: { body: {id: string; approved: boolean; comments: string;} }) => {
      const { promise, cancel } = resolver.putWorkflowApproval(args);
      if (cancelRequestRef?.current) {
        cancelRequestRef.current = cancel;
      }
      return promise;
    },
    {
      onSuccess: () => {
        queryCache.invalidateQueries(serviceUrl.getWorkflowExecution({ executionId }));
      },
    }
  );

  const handleSubmit = async (values: any) => {
    const body = {
      id: approvalId,
      approved: values.status === GateStatus.Approved,
      comments: values.comment,
    };
    try {
      await approvalMutator({ body });
      notify(
        <ToastNotification
          kind="success"
          title="Manual Approval"
          subtitle={body.approved ? "Successfully submitted approval request" : "Successfully submitted denial request"}
        />
      );
      closeModal();
    } catch (err) {
      // noop
    }
  };

  const buttons = [
    { icon: ThumbsDown16, label: "Reject", type: "negative", value: GateStatus.Rejected },
    { icon: ThumbsUp16, label: "Approve", type: "positive", value: GateStatus.Approved },
  ];

  return (
    <Formik
      enableReinitialize
      initialValues={{ comment: "", status: "" }}
      onSubmit={handleSubmit}
      validationSchema={Yup.object().shape({
        comment: Yup.string().nullable().max(200, "The comment must not have more than 200 characters"),
        status: Yup.string().nullable(),
      })}
    >
      {(props) => {
        const { values, handleSubmit, isValid, errors, touched, handleChange, setFieldValue, handleBlur } = props;

        const isApprovalApprovedOrRejected = Boolean(values.status);

        return (
          <ModalForm onSubmit={handleSubmit}>
            <ModalBody className={styles.modalBody}>
              {approvalsIsLoading ? (
                <Loading />
              ) : (
                <section className={styles.approval}>
                  {/*<time className={styles.approvalTime}>{`Submitted ${moment(creationDate).fromNow()}`}</time>*/}
                  <div className={styles.inputs}>
                    <div className={styles.comment}>
                      <TextArea
                        id={`comment`}
                        className={styles.commentArea}
                        labelText="Comments (optional)"
                        placeholder="Add some reasoning for your decision"
                        value={values?.comment}
                        onChange={handleChange}
                        onBlur={handleBlur}
                        invalid={errors?.comment && touched?.comment}
                        invalidText={errors?.comment}
                      />
                      <p className={styles.commentLength}>{`${values?.comment.length}/200`}</p>
                    </div>
                    <DecisionButtons
                      canUncheck
                      className={styles.decisionButtons}
                      items={buttons}
                      name={`decision-buttons`}
                      onChange={(value: string) => setFieldValue(`status`, value)}
                      selectedItem={values?.status}
                    />
                  </div>
                </section>
              )}

              {Boolean(approvalsError) && (
                <InlineNotification
                  style={{ marginBottom: "0.5rem" }}
                  lowContrast
                  kind="error"
                  title={"Manual Approval Failed"}
                  subtitle={"Something's Wrong"}
                />
              )}
            </ModalBody>
            <ModalFooter>
              <Button kind="secondary" type="button" onClick={closeModal}>
                Cancel
              </Button>
              <Button disabled={!isValid || approvalsIsLoading || !isApprovalApprovedOrRejected} type="submit">
                {!approvalsIsLoading ? "Submit decisions" : "Submitting..."}
              </Button>
            </ModalFooter>
          </ModalForm>
        );
      }}
    </Formik>
  );
}

export default TaskApprovalModal;
