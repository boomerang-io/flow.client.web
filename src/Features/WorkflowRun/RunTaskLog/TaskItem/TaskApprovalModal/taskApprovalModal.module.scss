.modalContainer {
  max-height: 51rem;
}

.modalBody {
  padding-right: 1rem;
}

.approval {
  border-top: 0.0625rem solid var(--cds-skeleton-02);
  padding: 1rem 0 0.5rem 0;
  position: relative;
  width: 100%;
}

.approvalTime {
  color: var(--cds-text-02);
  font-size: 0.75rem;
  letter-spacing: 0.02rem;
  line-height: 1rem;
  position: absolute;
  top: 1rem;
  right: 0rem;
}

.componentData {
  display: flex;
  margin-bottom: 1.5rem;
}

.pipelineData {
  display: flex;
  margin-bottom: 1rem;
}

.leftData {
  width: 12.6875rem;
}

.inputs {
  align-items: flex-end;
  display: flex;
}

.comment {
  position: relative;
  width: 37.75rem;
}

.commentArea {
  height: 4.5rem;
  resize: vertical !important;
}

.decisionButtons {
  margin: 0 0 1rem 1.5rem;
}
