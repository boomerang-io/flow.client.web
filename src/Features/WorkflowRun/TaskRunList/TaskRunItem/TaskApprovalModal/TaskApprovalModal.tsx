import {
  DecisionButtons,
  Loading,
  ModalForm,
  notify,
  TextArea,
  ToastNotification,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { Button, InlineNotification, ModalBody, ModalFooter } from "@carbon/react";
import { ThumbsUp, ThumbsDown } from "@carbon/react/icons";
import { useQueryClient, useMutation } from "react-query";
import { Formik } from "formik";
import * as Yup from "yup";
import styles from "./taskApprovalModal.module.scss";
import { serviceUrl, resolver } from "Config/servicesConfig";

const GateStatus = {
  Approved: "APPROVED",
  Rejected: "REJECTED",
} as const;

type Props = {
  actionId?: string;
  closeModal: () => void;
  workflowRunId: string;
};

function TaskApprovalModal({ actionId, closeModal, workflowRunId }: Props) {
  const queryClient = useQueryClient();

  const {
    mutateAsync: approvalMutator,
    isLoading: approvalsIsLoading,
    error: approvalsError,
  } = useMutation(resolver.putAction, {
    onSuccess: () => {
      queryClient.invalidateQueries(serviceUrl.getWorkflowRun({ id: workflowRunId }));
    },
  });

  const handleSubmit = async (values: { status: string; comment: string }) => {
    const body = [
      {
        id: actionId,
        approved: values.status === GateStatus.Approved,
        comments: values.comment,
      },
    ];
    try {
      await approvalMutator({ body });
      notify(
        <ToastNotification
          kind="success"
          title="Manual Approval"
          subtitle={
            body[0].approved ? "Successfully submitted approval request" : "Successfully submitted denial request"
          }
        />,
      );
      closeModal();
    } catch (err) {
      // noop
    }
  };

  const buttons = [
    { icon: ThumbsDown, label: "Reject", type: "negative", value: GateStatus.Rejected },
    { icon: ThumbsUp, label: "Approve", type: "positive", value: GateStatus.Approved },
  ];

  return (
    <Formik
      enableReinitialize
      initialValues={{ comment: "", status: "" }}
      onSubmit={handleSubmit}
      validationSchema={Yup.object().shape({
        comment: Yup.string().nullable().max(200, "The comment must not have more than 200 characters"),
        status: Yup.string().nullable(),
      })}
    >
      {(props) => {
        const { values, handleSubmit, isValid, errors, touched, handleChange, setFieldValue, handleBlur } = props;

        const isApprovalApprovedOrRejected = Boolean(values.status);

        return (
          <ModalForm onSubmit={handleSubmit}>
            <ModalBody className={styles.modalBody}>
              {approvalsIsLoading ? (
                <Loading />
              ) : (
                <section className={styles.approval}>
                  <div className={styles.inputs}>
                    <div className={styles.comment}>
                      <TextArea
                        enableCounter
                        id={`comment`}
                        className={styles.commentArea}
                        labelText="Comments (optional)"
                        placeholder="Add some reasoning for your decision"
                        value={values?.comment}
                        onChange={handleChange}
                        onBlur={handleBlur}
                        invalid={Boolean(errors?.comment && touched?.comment)}
                        invalidText={errors?.comment}
                        maxCount={200}
                      />
                    </div>
                    <DecisionButtons
                      canUncheck
                      className={styles.decisionButtons}
                      items={buttons}
                      name={"decision-buttons"}
                      onChange={(value: string) => setFieldValue("status", value)}
                      selectedItem={values?.status}
                    />
                  </div>
                </section>
              )}
              {Boolean(approvalsError) && (
                <InlineNotification
                  style={{ marginBottom: "0.5rem" }}
                  lowContrast
                  kind="error"
                  title={"Manual Approval Failed"}
                  subtitle={"Something's Wrong"}
                />
              )}
            </ModalBody>
            <ModalFooter>
              <Button kind="secondary" type="button" onClick={closeModal}>
                Cancel
              </Button>
              <Button disabled={!isValid || approvalsIsLoading || !isApprovalApprovedOrRejected} type="submit">
                {!approvalsIsLoading ? "Submit" : "Submitting..."}
              </Button>
            </ModalFooter>
          </ModalForm>
        );
      }}
    </Formik>
  );
}

export default TaskApprovalModal;
