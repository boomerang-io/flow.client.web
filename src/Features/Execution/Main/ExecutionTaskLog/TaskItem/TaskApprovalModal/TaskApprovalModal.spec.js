import React from "react";
import TaskApprovalModal from "./index";
import { fireEvent } from "@testing-library/react";

const props = {
  approvalId: "1",
  executionId: "1",
  closeModal: () => {},
};

describe("TaskApprovalModal --- Snapshot", () => {
  it("Capturing Snapshot of TaskApprovalModal", () => {
    const { baseElement } = rtlRender(<TaskApprovalModal {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});

describe("TaskApprovalModal --- RTL", () => {
  it("Formik form submission state", async () => {
    const { getByText } = rtlRender(<TaskApprovalModal {...props} />);

    const submissionButton = getByText("Submit decisions");
    expect(submissionButton).toBeDisabled();
    const approvalButton = getByText("Approve");
    fireEvent.click(approvalButton);
    expect(submissionButton).toBeEnabled();
  });
});
