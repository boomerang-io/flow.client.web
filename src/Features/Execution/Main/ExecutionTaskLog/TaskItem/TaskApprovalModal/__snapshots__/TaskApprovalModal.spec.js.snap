// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskApprovalModal --- Snapshot Capturing Snapshot of TaskApprovalModal 1`] = `
<body
  id="app"
>
  <div>
    <form
      class="bx--bmrg-modal-form"
    >
      <div
        class="bx--modal-content modalBody"
      >
        <section
          class="approval"
        >
          <div
            class="inputs"
          >
            <div
              class="comment"
            >
              <div
                class="bx--bmrg-text-area"
              >
                <div
                  class="bx--form-item"
                >
                  <label
                    class="bx--label"
                    for="comment"
                  >
                    <div
                      style="display: flex;"
                    >
                      <div>
                        Comments (optional)
                      </div>
                    </div>
                  </label>
                  <div
                    class="bx--text-area__wrapper"
                  >
                    <textarea
                      aria-placeholder="Add some reasoning for your decision"
                      class="bx--text-area commentArea"
                      cols="50"
                      id="comment"
                      placeholder="Add some reasoning for your decision"
                      rows="4"
                    />
                  </div>
                </div>
              </div>
              <p
                class="commentLength"
              >
                0/200
              </p>
            </div>
            <div
              class="bx--bmrg-decision-buttons"
            >
              <div
                class="bx--form-item"
              >
                <div
                  class="bx--radio-button-group decisionButtons"
                >
                  <label
                    class="bx--bmrg-decision-button-wrapper"
                    for="decision-buttons-REJECTED-0"
                  >
                    <input
                      id="decision-buttons-REJECTED-0"
                      name="decision-buttons"
                      readonly=""
                      type="radio"
                      value="REJECTED"
                    />
                    <div
                      class="bx--bmrg-decision-button bx--bmrg-decision-button--negative"
                    >
                      <span>
                        Reject
                      </span>
                      <svg
                        aria-hidden="true"
                        class="bx--bmrg-decision-button-icon"
                        fill="currentColor"
                        focusable="false"
                        height="16"
                        preserveAspectRatio="xMidYMid meet"
                        viewBox="0 0 32 32"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M26.76,21A3.64,3.64,0,0,0,30,17V11c0-3.65-1.56-8-9-8H2V17H8.46L12,22.37,13,28.71v.12A1.51,1.51,0,0,0,14.43,30H16a3.22,3.22,0,0,0,.76,0A3.64,3.64,0,0,0,20,26V21h6A3.22,3.22,0,0,0,26.76,21ZM8,15H4V5H8Zm20,2c0,1.44-.56,2-2,2H18v7c0,1.44-.56,2-2,2H14.87l-.14-1L14,21.63,10,15.7V5H21c4.78,0,7,1.91,7,6Z"
                        />
                      </svg>
                    </div>
                  </label>
                  <label
                    class="bx--bmrg-decision-button-wrapper"
                    for="decision-buttons-APPROVED-1"
                  >
                    <input
                      id="decision-buttons-APPROVED-1"
                      name="decision-buttons"
                      readonly=""
                      type="radio"
                      value="APPROVED"
                    />
                    <div
                      class="bx--bmrg-decision-button bx--bmrg-decision-button--positive"
                    >
                      <span>
                        Approve
                      </span>
                      <svg
                        aria-hidden="true"
                        class="bx--bmrg-decision-button-icon"
                        fill="currentColor"
                        focusable="false"
                        height="16"
                        preserveAspectRatio="xMidYMid meet"
                        viewBox="0 0 32 32"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M26,11H20V6a3.64,3.64,0,0,0-3.24-4A3.22,3.22,0,0,0,16,2H14.43A1.51,1.51,0,0,0,13,3.17v.12L12,9.63,8.46,15H2V29H21c7.44,0,9-4.35,9-8V15a3.64,3.64,0,0,0-3.24-4A3.22,3.22,0,0,0,26,11ZM8,27H4V17H8Zm20-6c0,4.09-2.22,6-7,6H10V16.3l4-5.93L14.73,5l.14-1H16c1.44,0,2,.56,2,2v7h8c1.44,0,2,.56,2,2Z"
                        />
                      </svg>
                    </div>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
      <div
        class="bx--modal-footer bx--btn-set"
      >
        <button
          class="bx--btn bx--btn--secondary"
          tabindex="0"
          type="button"
        >
          Cancel
        </button>
        <button
          class="bx--btn bx--btn--primary bx--btn--disabled"
          disabled=""
          tabindex="0"
          type="submit"
        >
          Submit decisions
        </button>
      </div>
    </form>
  </div>
</body>
`;
